Source/WTF/wtf/cocoa/WorkgroupInterval.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WorkgroupInterval.h"
+
+namespace WTF {
+
+WorkgroupInterval::WorkgroupInterval(const char* name)
+    : m_interval(createInterval(name))
+{
+}
+
+work_interval_t WorkgroupInterval::createInterval(const char*)
+{
+    work_interval_t workIntervalHandle = nullptr;
+    const int rv = work_interval_create(&workIntervalHandle, WORK_INTERVAL_TYPE_CA_CLIENT | WORK_INTERVAL_FLAG_JOINABLE);
+    RELEASE_ASSERT(!rv);
+    RELEASE_ASSERT(workIntervalHandle);
+    WTFLogAlways("Workgroup interval = %p", workIntervalHandle);
+    return workIntervalHandle;
+}
+
+bool WorkgroupInterval::join()
+{
+    Locker locker { m_lock };
+    auto rv = work_interval_join(m_interval);
+    WTFLogAlways("Workgroup join = %d", rv);
+    RELEASE_ASSERT(!rv || rv == EALREADY);
+    return !rv;
+}
+
+bool WorkgroupInterval::leave()
+{
+    Locker locker { m_lock };
+    auto rv = work_interval_leave();
+    WTFLogAlways("Workgroup leave %d", rv);
+    return !rv;
+}
+
+bool WorkgroupInterval::start(uint64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds)
+{
+    Locker locker { m_lock };
+    if (!m_instance)
+        m_instance = work_interval_instance_alloc(m_interval);
```

Comment:
Should we just allocate this in our constructor?

Reply:
Fixed!